https://support.process-one.net/browse/EJAB-1170

diff --git a/src/Makefile.in b/src/Makefile.in
index 3be72ce..10261e9 100644
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -239,9 +239,9 @@ install: all
 	#
 	# Documentation
 	install -d $(DOCDIR)
-	install ../doc/guide.html $(DOCDIR)
-	install ../doc/*.png $(DOCDIR)
-	install ../doc/*.txt $(DOCDIR)
+	install -m 644 ../doc/guide.html $(DOCDIR)
+	install -m 644 ../doc/*.png $(DOCDIR)
+	install -m 644 ../doc/*.txt $(DOCDIR)
 
 uninstall: uninstall-binary
 
