diff -urpN libxcrypt-3.0.2.orig/configure.in libxcrypt-3.0.2/configure.in
--- libxcrypt-3.0.2.orig/configure.in	2008-09-26 15:34:03.145412363 -0600
+++ libxcrypt-3.0.2/configure.in	2008-09-26 15:34:43.144619919 -0600
@@ -29,7 +29,7 @@ AC_SYS_LARGEFILE
 
 EXTRA_CFLAGS=""
 if eval "test x$GCC = xyes"; then
-    EXTRA_CFLAGS="-W -Wall -Wbad-function-cast -Wcast-align -Wcast-qual -Winline -Wmissing-declarations -Wmissing-prototypes -Wnested-externs -Wshadow -Wstrict-prototypes -Wundef -Wpointer-arith -Werror"
+    EXTRA_CFLAGS="-W -Wall -Wbad-function-cast -Wcast-align -Wcast-qual -Winline -Wmissing-declarations -Wmissing-prototypes -Wnested-externs -Wstrict-prototypes -Wundef -Wpointer-arith -Werror"
 fi
 AC_SUBST(EXTRA_CFLAGS)
 
