--- a/setup.py	2008-08-15 02:28:18.000000000 -0600
+++ b/setup.py	2009-02-24 11:02:47.229228029 -0700
@@ -10,5 +10,7 @@
        author_email = "awalters@volatilesystems.com",  
        url          = "http://www.volatilesystems.com",
        license      = "GPL",
-       packages     = ["forensics", "forensics.win32","memory_plugins","memory_objects","memory_objects.Linux","memory_objects.Windows","thirdparty"],
+       packages     = ["forensics", "forensics.win32","memory_plugins","memory_objects","memory_objects.Windows","thirdparty"],
+       py_modules   = ['vmodules', 'vsyms', 'vtypes', 'vutils'],
+       scripts      = ['volatility'],
        )
