diff --git a/setup.py b/setup.py
index e9c0641..2fc2acc 100755
--- a/setup.py
+++ b/setup.py
@@ -68,7 +68,7 @@ setup(
         'Operating System :: OS Independent',
         'Programming Language :: Python',
     ],
-    packages=find_packages('.', exclude=[u'examples', u'tests']),
+    packages=find_packages('.', exclude=[u'examples', u'tests', u'tests.*']),
     package_dir={
         'dfvfs': 'dfvfs'
     },
