diff -urpN a/acinclude.m4 b/acinclude.m4
--- a/acinclude.m4	2009-03-10 14:47:52.000000000 -0600
+++ b/acinclude.m4	2009-05-06 10:40:27.066276912 -0600
@@ -234,9 +234,9 @@ AC_DEFUN([LIBEWF_CHECK_PYTHON_DEVEL],
 	AC_MSG_CHECKING(
 	 [for Python library path])
 
-	PYTHON_LDFLAGS=`$PYTHON -c "import distutils.sysconfig;print distutils.sysconfig.get_python_lib() "`;
+	python_path=`$PYTHON -c "import distutils.sysconfig;print distutils.sysconfig.get_python_lib() "`;
 	AC_MSG_RESULT(
-	 [$PYTHON_LDFLAGS])
+	 [$python_path])
 
 	AC_SUBST(
 	 [PYTHON_LDFLAGS],
diff -urpN a/ewftools/alteration_handle.c b/ewftools/alteration_handle.c
--- a/ewftools/alteration_handle.c	2009-05-05 11:54:50.000000000 -0600
+++ b/ewftools/alteration_handle.c	2009-05-06 10:40:20.945983762 -0600
@@ -504,7 +504,7 @@ ssize_t alteration_handle_prepare_read_b
 	                 storage_media_buffer->process_crc,
 	                 error );
 #else
-	process_count = libewf_raw_prepare_read_buffer(
+	process_count = libewf_raw_read_prepare_buffer(
 	                 alteration_handle->input_handle,
 	                 storage_media_buffer->compression_buffer,
 	                 storage_media_buffer->compression_buffer_amount,
@@ -678,7 +678,7 @@ ssize_t alteration_handle_prepare_write_
 	                 &( storage_media_buffer->process_crc ),
 	                 error );
 #else
-	process_count = libewf_raw_prepare_write_buffer(
+	process_count = libewf_raw_write_prepare_buffer(
 	                 alteration_handle->input_handle,
 	                 storage_media_buffer->raw_buffer,
 	                 storage_media_buffer->raw_buffer_amount,
diff -urpN a/ewftools/export_handle.c b/ewftools/export_handle.c
--- a/ewftools/export_handle.c	2009-05-05 11:53:46.000000000 -0600
+++ b/ewftools/export_handle.c	2009-05-06 10:40:20.947983807 -0600
@@ -809,7 +809,7 @@ ssize_t export_handle_prepare_read_buffe
 	                 storage_media_buffer->process_crc,
 	                 error );
 #else
-	process_count = libewf_raw_prepare_read_buffer(
+	process_count = libewf_raw_read_prepare_buffer(
 	                 export_handle->input_handle,
 	                 storage_media_buffer->compression_buffer,
 	                 storage_media_buffer->compression_buffer_amount,
@@ -1076,7 +1076,7 @@ ssize_t export_handle_prepare_write_buff
 				 &( storage_media_buffer->process_crc ),
 				 error );
 #else
-		process_count = libewf_raw_prepare_write_buffer(
+		process_count = libewf_raw_write_prepare_buffer(
 				 export_handle->ewf_output_handle,
 				 storage_media_buffer->raw_buffer,
 				 storage_media_buffer->raw_buffer_amount,
diff -urpN a/ewftools/imaging_handle.c b/ewftools/imaging_handle.c
--- a/ewftools/imaging_handle.c	2009-05-05 11:52:33.000000000 -0600
+++ b/ewftools/imaging_handle.c	2009-05-06 10:40:20.950993758 -0600
@@ -576,7 +576,7 @@ ssize_t imaging_handle_prepare_read_buff
 	                 storage_media_buffer->process_crc,
 	                 error );
 #else
-	process_count = libewf_raw_prepare_read_buffer(
+	process_count = libewf_raw_read_prepare_buffer(
 	                 imaging_handle->output_handle,
 	                 storage_media_buffer->compression_buffer,
 	                 storage_media_buffer->compression_buffer_amount,
@@ -770,7 +770,7 @@ ssize_t imaging_handle_prepare_write_buf
 	                 &( storage_media_buffer->process_crc ),
 	                 error );
 #else
-	process_count = libewf_raw_prepare_write_buffer(
+	process_count = libewf_raw_write_prepare_buffer(
 	                 imaging_handle->output_handle,
 	                 storage_media_buffer->raw_buffer,
 	                 storage_media_buffer->raw_buffer_amount,
diff -urpN a/ewftools/verification_handle.c b/ewftools/verification_handle.c
--- a/ewftools/verification_handle.c	2009-05-05 11:50:44.000000000 -0600
+++ b/ewftools/verification_handle.c	2009-05-06 10:40:20.951983549 -0600
@@ -568,7 +568,7 @@ ssize_t verification_handle_prepare_read
 	                 storage_media_buffer->process_crc,
 	                 error );
 #else
-	process_count = libewf_raw_prepare_read_buffer(
+	process_count = libewf_raw_read_prepare_buffer(
 	                 verification_handle->input_handle,
 	                 storage_media_buffer->compression_buffer,
 	                 storage_media_buffer->compression_buffer_amount,
diff -urpN a/pyewf/pyewf_file.c b/pyewf/pyewf_file.c
--- a/pyewf/pyewf_file.c	2009-04-19 02:55:34.000000000 -0600
+++ b/pyewf/pyewf_file.c	2009-05-06 10:40:37.165983905 -0600
@@ -1,4 +1,4 @@
-
+/*
  * File Object definition for libewf Python bindings
  *
  * Copyright (c) 2008, David Collett <david.collett@gmail.com>
