--- makesnortsam.sh.orig	2005-01-25 07:49:43.432290588 -0500
+++ makesnortsam.sh	2005-01-25 07:51:53.722343371 -0500
@@ -13,15 +13,13 @@
 
 #BSDTHREADLIB='-lc_r'
 BSDTHREADLIB='-lpthread'
-
+source ${ROOT}sbin/functions.sh
 systype=`uname`
 
 case "$1" in
 	[oO][pP][sS][eE][cC])
 
-		echo "-------------------------------------------------------------------------------"
-		echo "Building SnortSam (release) with OPSEC libraries"
-		echo "-------------------------------------------------------------------------------"
+		einfo "Building SnortSam (release) with OPSEC libraries"
 
 		rm -f snortsam
 		rm -f snortsam-debug
@@ -60,9 +58,7 @@
 		esac
 
 
-		echo "-------------------------------------------------------------------------------"
-		echo "Building SnortSam (debug) with OPSEC libraries"
-		echo "-------------------------------------------------------------------------------"
+		einfo "Building SnortSam (debug) with OPSEC libraries"
 
 		rm -f *.o
 
@@ -99,9 +95,7 @@
 		cd ..
 		;;
 	*)
-		echo "-------------------------------------------------------------------------------"
-		echo "Building SnortSam (release)"
-		echo "-------------------------------------------------------------------------------"
+		einfo "Building SnortSam (release)"
 
 		rm -f snortsam
 		rm -f snortsam-debug
@@ -140,9 +134,7 @@
 		esac
 
 
-		echo "-------------------------------------------------------------------------------"
-		echo "Building SnortSam (debug)"
-		echo "-------------------------------------------------------------------------------"
+		einfo "Building SnortSam (debug)"
 
 		rm -f *.o
 
